File: 2020/b.py
#!/usr/bin/env python3

# Problem B - Contact Tracer

def solve(p, ab):
    infect = {p}
    for a, b in ab:
        if a in infect or b in infect:
            infect.update((a, b))
    return len(infect)

while True:
    m, n, p = map(int, input().split())
    if m == 0:
        exit()
    ab = []
    for i in range(n):
        a, b = map(int, input().split())
        ab.append((a, b))
    print(solve(p, ab))

File: 2020/b.hs
#!/usr/bin/env runghc

-- ACM-ICPC 2020 Domestic Qualifier
-- Problem B: Contact Tracer

import Control.Monad (when, replicateM)
import Control.Applicative ((<$>))
import qualified Data.Set as Set

main :: IO ()
main = do
    [m, n, p] <- map read . words <$> getLine :: IO [Int]
    when ((m, n, p) /= (0, 0, 0)) $ do
        ab <- replicateM n (map read . words <$> getLine) :: IO [[Int]]
        print $ solve p ab
        main

solve :: Int -> [[Int]] -> Int
solve p = Set.size . foldl updater (Set.singleton p)
    where updater set [a, b] =
              if Set.member a set then Set.insert b set else
              if Set.member b set then Set.insert a set else set
